cranelift/assembler-x64/meta/src/dsl/encoding.rs

```diff
@@ -301,117 +298,256 @@ pub struct Opcodes {
 impl From<u8> for Opcodes {
     fn from(primary: u8) -> Opcodes {
         Opcodes {
-            prefix: LegacyPrefix::NoPrefix,
+            prefixes: Prefixes::default(),
             escape: false,
             primary,
             secondary: None,
         }
     }
 }
 
-impl From<[u8; 1]> for Opcodes {
-    fn from(bytes: [u8; 1]) -> Opcodes {
-        Opcodes::from(bytes[0])
+impl<const N: usize> From<[u8; N]> for Opcodes {
+    fn from(bytes: [u8; N]) -> Self {
+        let (prefixes, remaining) = Prefixes::parse(&bytes);
+        let (escape, primary, secondary) = match remaining {
+            [primary] => (false, *primary, None),
+            [0x0f, primary] => (true, *primary, None),
+            [0x0f, primary, secondary] => (true, *primary, Some(*secondary)),
+            _ => panic!("invalid opcodes after prefix; expected [opcode], [0x0f, opcode], or [0x0f, opcode, opcode], found {remaining:?}"),
+        };
+        Self { prefixes, escape, primary, secondary }
     }
 }
 
-impl From<[u8; 2]> for Opcodes {
-    fn from(bytes: [u8; 2]) -> Opcodes {
-        let [a, b] = bytes;
-        match (LegacyPrefix::try_from(a), b) {
-            (Ok(prefix), primary) => Opcodes { prefix, escape: false, primary, secondary: None },
-            (Err(0x0f), primary) => Opcodes {
-                prefix: LegacyPrefix::NoPrefix,
-                escape: true,
-                primary,
-                secondary: None,
-            },
-            _ => panic!("invalid opcodes; expected [prefix, opcode] or [0x0f, opcode]"),
+/// The allowed prefixes for an instruction. From the reference manual (section
+/// 2.1.1):
+///
+/// > Instruction prefixes are divided into four groups, each with a set of
+/// > allowable prefix codes. For each instruction, it is only useful to include
+/// > up to one prefix code from each of the four groups (Groups 1, 2, 3, 4).
+/// > Groups 1 through 4 may be placed in any order relative to each other.
+#[derive(Default)]
+pub struct Prefixes {
+    pub group1: Option<Group1Prefix>,
+    pub group2: Option<Group2Prefix>,
+    pub group3: Option<Group3Prefix>,
+    pub group4: Option<Group4Prefix>,
+}
+
+impl Prefixes {
+    /// Parse a slice of `bytes` into a set of prefixes, returning both the
+    /// configured [`Prefixes`] as well as any remaining bytes.
+    fn parse(mut bytes: &[u8]) -> (Self, &[u8]) {
+        let mut prefixes = Self::default();
+        while prefixes.try_assign(bytes[0]).is_ok() {
```

Comment:
`while bytes.len() > 0 && prefixes.try_assign(bytes[0]).is_ok()` ?

(I think this won't matter as long as you only pass in slices that end in true opcodes, and no opcode has the same value as a prefix, but it caught my eye as that's quite subtle)